rfd/0080-yubikey-piv.md

```diff
@@ -0,0 +1,272 @@
+spec:
+  yubikey_login: 
+    required: false
+    pin_policy: once
+    touch_policy: cached
+---
+kind: role
+version: v5
+metadata:
+  name: yubikey-pin
+spec:
+  yubikey_login: 
+    required: true
+    pin_policy: always
+---
+kind: role
+version: v5
+metadata:
+  name: yubikey-touch
+spec:
+  yubikey_login: 
+    required: true
+    pin_policy: never
+```
+
+### UX
+
+#### yubikey slots
+
+Yubikey 5 has 5 different [key slots](https://docs.yubico.com/hardware/yubikey/yk-5/tech-manual/yk5-apps.html#slot-information) to choose from, one of which is reserved for attestation. `Slot 9a: PIV Authentication` is the most fitting option, and is the same slot used by `yubikey-agent`.
+
+If the slot is already in use, we will provide the user with a prompt to reuse this key, overwrite it, or abort. Users can also provide the `--yubikey-overwrite` or `--yubikey-reuse` flags to bypass this prompt.
+
+```
+> tsh login
+
+yubikey slot 9a has a pre-generated private key with public key:
+-----BEGIN PUBLIC KEY-----
+...
+-----END PUBLIC KEY-----
+
+Would you like to:
+ (1) Resuse this key
+ (2) Overwrite this key
+ (3) Abort login
+```
+
+In the case of a Teleport user connecting to multiple clusters, they'll need to use `--yubikey-reuse` in order to maintain access to them simultaneously.
+
+Note that reusing the key is not any less secure than generating a new one, as the Teleport Auth server can still attest that the key was generated on the yubikey, and therefore cannot be exported. 
+
+#### new `tsh` flags
+
+| Name | Default Value(s) | Allowed Value(s) | Description |
+| `-y, --yubikey` | - | - | During login, use yubikey to generate and store private key data for your login session. |
+| `--yubikey-pin-policy` | never | always once never | Can be provided with `tsh login -y` to control how often you need to enter your yubikey PIN to access the yubikey private key data. |
+| `--yubikey-touch-policy` | never | always cached never | Can be provided with `tsh login -y` to control how often you need to touch your yubikey to access the yubikey private key data. |
+| `--yubikey-overwrite` | - | - | When provided, `tsh login -y` will overwrite existing yubikey data if it exists. |
+| `--yubikey-reuse` | - | - | When provided, `tsh login -y` will reuse existing yubikey data if it exists. Useful when using yubikey to connect to multiple clusters simultaneously. |
+
+#### tsh config file
+
+Users can also provide their yubikey login settings in their tsh config file (`/etc/tsh.yaml` or `~/.tsh/config/config.yaml`). This is important so that users don't need to provide the flag options on every login.
+
+```yaml
+yubikey: true
+yubikey_pin: once
+yubikey_touch: cached
+yubikey_reuse: true
+```
+
+#### Multiple yubikeys
+
+If the user has multiple yubikeys connected, they will receive an error asking them to connect just one yubikey. Otherwise, we can't guarentee that the right yubikey will be chosen. We may be able to add support for this in the future.
```

Comment:
Where does that limitation come from? Is it inherent to the protocol, Go APIs, etc?

Reply:
I'm just not sure if there's a good way to do this right now. We could give the user a prompt like:
```
> tsh login --yubikey-piv
multiple yubikey devices detected, please choose one from the list:
 (1) YubiKey 5 Nano (5.2.7) [OTP+FIDO+CCID] Serial: xxxxxxxx
 (2) YubiKey 5 Nano (5.1.1) [OTP+FIDO] Serial: xxxxxxxx
```

but I'm not sure there's a way to choose one by touch like with webauthn. Perhaps it's a limitation with the protocol, or maybe just `go-piv` :disappointed: I'll look into this

Reply:
Since we are constrained to Yubikeys we could use a mock WebAuthn assertion to choose the device by touch, then it's a matter of trying to find the same device via the PIV API. One could argue that this is too clever, or crosses the streams too much, which I would concede.

Another option is a --yubikey-piv-serial flag. If we keep tabs of device serials that have PIV keys in them, then we'd only need the user to pass the flag once. (Just throwing out some ideas.)